libs/data/bbhnet/data/dataloader.py

```diff
@@ -0,0 +1,449 @@
+            # way to make sure wer'e slicing from the
+            # last dimension
+            if len(array.shape) == 2:
+                samples.append(array[i, start:stop])
+            else:
+                samples.append(array[i, :, start:stop])
+
+        # stack the samples into a tensor
+        return torch.stack(samples)
+
+    def sample_from_background(self, independent: bool = True):
+        """Sample a batch of kernels from the background data
+        Randomly sample kernels from the interferometer
+        background timeseries in a uniform manner. If
+        `independent == True`, kernels will be sampled
+        from each interferometer independently and then
+        concatenate to simulate fully random time-shifting.
+        """
+
+        # sample kernel start indices uniformly
+        # from the hanford background data ensuring
+        # that the corresponding stop index will fall
+        # within the timeseries
+        hanford_start = np.random.choice(
+            len(self.hanford_background) - self.kernel_size,
+            size=self.batch_size,
+            replace=False,
+        )
+
+        # if sampling indepdently, create a second set
+        # of start indices for the livingston background
+        # data. Otherwise, use the same indices as for hanford
+        if independent:
+            livingston_start = np.random.choice(
+                len(self.livingston_background) - self.kernel_size,
+                size=self.batch_size,
+                replace=False,
+            )
+        else:
+            livingston_start = hanford_start
+
+        # grab kernels for each start index from the background
+        # TODO: is there a good way to do this with array ops?
+        X = []
+        for h_idx, l_idx in zip(hanford_start, livingston_start):
+            hanford = self.hanford_background[h_idx : h_idx + self.kernel_size]
+            livingston = self.livingston_background[
+                l_idx : l_idx + self.kernel_size
+            ]
+            X.extend([hanford, livingston])
+
+        # stack everything to make a (batch_size * 2, sample_rate)
+        # sized tensor, then reshape to move the two
+        # interferometers to the 1st dimension
+        X = torch.stack(X, dim=0).reshape(self.batch_size, 2, -1)
+        return X
+
+    def _sample_ra(self, num_samples):
+        raise NotImplementedError
+
+    def _sample_dec(self, num_samples):
+        raise NotImplementedError
+
+    def _sample_psi(self, num_samples):
+        raise NotImplementedError
+
+    def _sample_geocent_time(self, num_samples):
+        raise NotImplementedError
+
+    def inject_waveforms(self, background, waveforms):
+        sample_params = {
+            "ra": self._sample_ra(len(background)),
+            "dec": self._sample_dec(len(background)),
+            "psi": self._sample_psi(len(background)),
+            "geocent_time": self._sample_geocent_time(len(background)),
+        }

```

Comment:
@wbenoit26 any ideas as to what these sampling functions might be? I assume at least one of them is like a uniform [0, pi] sample or something but not familiar enough with the meaning of all the parameters to be sure

Reply:
We have some freedom to do what we want here, but I think it makes sense to stick with the priors from some of bilby's example prior dictionaries. They have a Uniform(0, 2*pi) prior on RA, a Cosine prior on Dec, and if the black hole is spinning, a Uniform(0, pi) prior on psi (0 if not spinning). We could have bilby do the sampling here too, but it may not be worth the overhead.

Geocent_time has just been a uniform on the duration of the frame, plus the initial time. For my injection code, there's some logic to prevent signals from overlapping.

Reply:
@wbenoit26 I opted to just use bilby's samplers since they were the most straightforward to implement. I may experiment with using scipy's distribution functions (since numpy doesn't support cosine), but this should work for now.